lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,667 @@
+// or tx_ack_rbf          │     │     │               │  │          ┌─────────────────┐ │                │                     │   │
+//    (sent or received)  │     │ ┌──►└───────────────┘  │          │                 │ │                └─────────────────────┘   │
+//                              │ │                      │          │ TheirTxComplete ├─┘                                          │
+//             sent_tx_complete │ │ received_tx_add_*    │          │                 │                   ┌────────────────────┐   │
+//                              │ │ received_tx_remove_* │          └─────────────────┘                   │                    │   │
+//                        │     │ │                      │                                            ┌──►│ NegotiationAborted │   │
+//                        │     │ └───┬───────────────┐  │        (sent/received)_tx_abort            │   │                    │   │
+//                        │     │     │               │  ├────────────────────────────────────────────┘   └────────────────────┘   │
+//                        │     └────►│ OurTxComplete │  │                                                                         │
+//                        │           │               ├──┼──┐                                                                      │
+//                        │           └───────────────┘  │  └──────────────────────────────────────────────────────────────────────┘
+//                        │                              │                         received_tx_complete
+//                        │                              │
+//                        └──────────────────────────────┘
+//
+
+// Channel states that can receive `(send|receive)_tx_(add|remove)_(input|output)`
+pub(crate) trait AcceptingChanges {}
+
+/// We are currently in the process of negotiating the transaction.
+pub(crate) struct Negotiating;
+/// We have sent a `tx_complete` message and are awaiting the counterparty's.
+pub(crate) struct OurTxComplete;
+/// We have received a `tx_complete` message and the counterparty is awaiting ours.
+pub(crate) struct TheirTxComplete;
+/// We have exchanged consecutive `tx_complete` messages with the counterparty and the transaction
+/// negotiation is complete.
+pub(crate) struct NegotiationComplete;
+/// The negotiation has failed and cannot be continued.
+pub(crate) struct NegotiationAborted(AbortReason);
+
+impl AcceptingChanges for Negotiating {}
+impl AcceptingChanges for OurTxComplete {}
+
+struct NegotiationContext {
+	channel_id: [u8; 32],
+	require_confirmed_inputs: bool,
+	holder_is_initiator: bool,
+	received_tx_add_input_count: u16,
+	received_tx_add_output_count: u16,
+	inputs: HashMap<u64, TxIn>,
+	prevtx_outpoints: HashSet<OutPoint>,
+	outputs: HashMap<u64, TxOut>,
+	base_tx: Transaction,
+	did_send_tx_signatures: bool,
```

Comment:
It doesn’t seem considered as a spec failure to re-send `tx_signatures` should we have something else than a boolean ? and it sounds a valid behavior if we’re negotiating rbf